### ISSUE TITLE
JSON serialization failure when agent initialized with a PosixPath object

### ISSUE THREAD

**Issue by ps-jay** [newrelic/newrelic-python-agent#16]
When initializing the New Relic agent with the configuration file provided as a PosixPath object, then the agent will fail with a JSON serialization failure.

Issue occurs on 5.18.0.148
Issue doesn't not occur on 5.14.0.142

**Description**
When initializing the agent like this:
```
from pathlib import Path
import newrelic.agent

BASE_DIR = Path(__file__).parents[1]
newrelic.agent.initialize(BASE_DIR / 'newrelic.ini')  # initialize with a `PosixPath` type
```
We get this failure:
```
{
"asctime": "2020-09-09 06:32:07,356",
"levelname": "ERROR",
"name": "newrelic.core.application",
"message": "Unexpected exception when registering agent with the data collector. If this problem persists, please report this problem to New Relic support for further investigation.",
"pathname": "/venv/lib/python3.8/site-packages/newrelic/core/application.py",
"lineno": 367,
"funcName": "connect_to_data_collector",
"exc_info": "Traceback (most recent call last):\n File \"/venv/lib/python3.8/site-packages/newrelic/core/application.py\", line 346, in connect_to_data_collector\n active_session = create_session(None, self._app_name,\n File \"/venv/lib/python3.8/site-packages/newrelic/core/data_collector.py\", line 232, in create_session\n return DeveloperModeSession(\n File \"/venv/lib/python3.8/site-packages/newrelic/core/data_collector.py\", line 40, in __init__\n self._protocol = self.PROTOCOL.connect(\n File \"/venv/lib/python3.8/site-packages/newrelic/core/agent_protocol.py\", line 458, in connect\n protocol.send(\"agent_settings\", (global_settings_dump(settings),))\n File \"/venv/lib/python3.8/site-packages/newrelic/core/agent_protocol.py\", line 205, in send\n params, headers, payload = self._to_http(method, payload)\n File \"/venv/lib/python3.8/site-packages/newrelic/core/agent_protocol.py\", line 251, in _to_http\n return params, self._headers, json_encode(payload).encode(\"utf-8\")\n File \"/venv/lib/python3.8/site-packages/newrelic/common/encoding_utils.py\", line 104, in json_encode\n return json.dumps(obj, **_kwargs)\n File \"/usr/local/lib/python3.8/json/__init__.py\", line 234, in dumps\n return cls(\n File \"/usr/local/lib/python3.8/json/encoder.py\", line 199, in encode\n chunks = self.iterencode(o, _one_shot=True)\n File \"/usr/local/lib/python3.8/json/encoder.py\", line 257, in iterencode\n return _iterencode(o, 0)\n File \"/venv/lib/python3.8/site-packages/newrelic/common/encoding_utils.py\", line 91, in _encode\n raise TypeError(repr(o) + ' is not JSON serializable')\nTypeError: PosixPath('/app/newrelic.ini') is not JSON serializable",
"correlation_id": "30d062bf-5d60-4fe5-9409-fdd7f52f532e",
"forwarded_user": null
}
```

**Expected Behavior**
Previously (version 5.14.0.142 or older) this form of initialization succeeded.

**Troubleshooting**
To investigate, I have altered the initialization to pass in a string, rather than a `PosixPath` object:
```
from pathlib import Path
import newrelic.agent

BASE_DIR = Path(__file__).parents[1]
newrelic.agent.initialize(f"{BASE_DIR / 'newrelic.ini'}")  # initialize with a `str` type
```

When doing this, the exception does not occur, and the agent initializes and connects successfully.

**Steps to Reproduce**
Initialize the 5.18.0.148 agent with a `PosixPath` object.

**Your Environment**
Python 3.8 + Django 2.2.16 running in a Docker container with the New Relic Python agent 5.18.0.148.

**Additional context**
New Relic support case [425788](https://support.newrelic.com/tickets/425788)

**Comment by robertknight** [comment]
This problem has started to reoccur with the transition to https://github.com/hypothesis/h/pull/9128 as a result of https://github.com/newrelic/newrelic-python-agent/commit/14a219fc93fca627e994cd5978f737deadb73d76.

Example trace from our app:

```
  websocket (stderr)   | AttributeError: 'PosixPath' object has no attribute 'endswith'
   websocket (stderr)   | 2024-11-27 16:37:21,920 [1066] [gunicorn.error:INFO] Worker exiting (pid: 1066)
   websocket (stderr)   | Traceback (most recent call last):
   websocket (stderr)   |   File "src/gevent/greenlet.py", line 900, in gevent._gevent_cgreenlet.Greenlet.run
   websocket (stderr)   |   File "/var/lib/hypothesis/h/streamer/metrics.py", line 52, in metrics_process
   websocket (stderr)   |     newrelic.agent.initialize(config_file=config_file)
   websocket (stderr)   |   File "/usr/local/lib/python3.11/site-packages/newrelic/config.py", line 4803, in initialize
   websocket (stderr)   |     _load_configuration(config_file, environment, ignore_errors, log_file, log_level)
   websocket (stderr)   |   File "/usr/local/lib/python3.11/site-packages/newrelic/config.py", line 1027, in _load_configuration
   websocket (stderr)   |     if config_file.endswith(".toml"):
   websocket (stderr)   |        ^^^^^^^^^^^^^^^^^^^^
   websocket (stderr)   | AttributeError: 'PosixPath' object has no attribute 'endswith'
```

The code that calls `newrelic.agent.initialize` is here: https://github.com/hypothesis/h/blob/70c33c8afaa395ecf8c792565a8b42f07c80f27b/h/streamer/metrics.py#L52

**Comment by lybekk** [comment]
Can confirm this applies to `10.3.1` as well

**Comment by hmstepanek** [comment]
This issue has been fixed in v10.4.0.
